File: ios/CalibraSample/CalibraSample/Sections/VocalRangeSection.swift
import SwiftUI
import vozos

/// Vocal Range Detector using VocalRangeSession API.
///
/// Demonstrates:
/// - Direct `VocalRangeSession` API with `@State` primitives (iOS 15+ pattern)
/// - `SonixRecorder` for audio capture
/// - `MainActor.run` for thread-safe UI updates
struct VocalRangeSection: View {
    // Direct Calibra API + @State primitives (recommended iOS 15+ pattern)
    @State private var session: VocalRangeSession?
    @State private var recorder: SonixRecorder?
    @State private var audioTask: Task<Void, Never>?
    @State private var observerTask: Task<Void, Never>?

    // State primitives from VocalRangeState
    @State private var phase: VocalRangePhase = .idle
    @State private var countdownSeconds: Int32 = 0
    @State private var phaseMessage: String = "Ready to detect your vocal range"
    @State private var currentPitch: VocalPitch? = nil
    @State private var currentAmplitude: Float = 0
    @State private var stabilityProgress: Float = 0
    @State private var bestLowNote: DetectedNote? = nil
    @State private var bestHighNote: DetectedNote? = nil
    @State private var lowNote: DetectedNote? = nil
    @State private var highNote: DetectedNote? = nil
    @State private var result: VocalRangeResult? = nil
    @State private var error: String? = nil

    // Computed properties
    private var currentPitchLabel: String {
        currentPitch?.noteLabel ?? "-"
    }

    private var currentPitchHz: Float {
        currentPitch?.frequencyHz ?? 0
    }

    private var isDetecting: Bool {
        phase == .detectingLow || phase == .detectingHigh
    }

    private var hasError: Bool {
        error != nil
    }

    /// Best note for current detection phase
    private var bestNoteForPhase: DetectedNote? {
        switch phase {
        case .detectingLow: return bestLowNote
        case .detectingHigh: return bestHighNote
        default: return nil
        }
    }

    /// Whether a Lock button should be shown
    private var canLockNote: Bool {
        bestNoteForPhase != nil
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Vocal Range Detector")
                .font(.headline)

            Text(error ?? phaseMessage)
                .font(.subheadline)
                .foregroundColor(hasError ? .red : .secondary)

            // Phase indicators
            phaseIndicators

            // Countdown
            if phase == .countdown && countdownSeconds > 0 {
                Text("\(countdownSeconds)")
                    .font(.system(size: 72, weight: .bold))
                    .frame(maxWidth: .infinity)
            }

            // Real-time pitch display during detection
            if isDetecting {
                pitchDisplayCard

                // Best note so far card
                if let best = bestNoteForPhase {
                    bestNoteCard(note: best)
                }

                // Stability progress
                VStack(alignment: .leading, spacing: 4) {
                    Text("Stability: Hold for 1 second")
                        .font(.caption)
                        .foregroundColor(.secondary)
                    ProgressView(value: Double(min(max(stabilityProgress, 0), 1)))
                        .tint(stabilityProgress >= 1.0 ? .green : .blue)
                }

                // Level meter
                HStack {
                    Text("Level:")
                        .font(.caption)
                    ProgressView(value: Double(min(max(currentAmplitude, 0), 1)))
                }
            }

            // Results
            if let result = result {
                resultsCard(result: result)
            }

            // Controls
            controlButtons
        }
        .onDisappear {
            stop()
            cleanup()
        }
    }

    /// Whether the flow is in a state where phases should show as complete
    private var isFlowActive: Bool {
        phase != .cancelled && phase != .idle
    }

    private var phaseIndicators: some View {
        HStack {
            Spacer()
            PhaseIndicator(
                label: "Ready",
                isActive: phase == .countdown,
                isComplete: isFlowActive && phase.ordinal > VocalRangePhase.countdown.ordinal
            )
            Text("→")
                .padding(.top, 8)
            PhaseIndicator(
                label: "Low",
                isActive: phase == .detectingLow,
                isComplete: isFlowActive && phase.ordinal > VocalRangePhase.detectingLow.ordinal
            )
            Text("→")
                .padding(.top, 8)
            PhaseIndicator(
                label: "High",
                isActive: phase == .detectingHigh,
                isComplete: phase == .complete
            )
            Text("→")
                .padding(.top, 8)
            PhaseIndicator(
                label: "Done",
                isActive: phase == .complete,
                isComplete: false
            )
            Spacer()
        }
    }

    private var pitchDisplayCard: some View {
        VStack(spacing: 4) {
            Text(currentPitchLabel)
                .font(.system(size: 48, weight: .bold))
            Text(String(format: "%.1f Hz", currentPitchHz))
                .font(.title3)
        }
        .frame(maxWidth: .infinity)
        .padding(16)
        .background(stabilityProgress >= 1.0 ? Color.green.opacity(0.3) : Color.blue.opacity(0.2))
        .cornerRadius(12)
    }

    private func bestNoteCard(note: DetectedNote) -> some View {
        let label = phase == .detectingLow ? "Lowest so far" : "Highest so far"
        return VStack(spacing: 8) {
            HStack {
                VStack(alignment: .leading, spacing: 2) {
                    Text(label)
                        .font(.caption)
                        .foregroundColor(.white.opacity(0.8))
                    HStack(alignment: .firstTextBaseline, spacing: 4) {
                        Text(note.pitch.noteLabel)
                            .font(.title2)
                            .fontWeight(.bold)
                            .foregroundColor(.white)
                        Text(String(format: "%.1f Hz", note.pitch.frequencyHz))
                            .font(.caption)
                            .foregroundColor(.white.opacity(0.7))
                    }
                }
                Spacer()
                Button("Lock & Continue") {
                    _ = session?.confirmNote()
                }
                .buttonStyle(.bordered)
                .tint(.white)
            }
        }
        .padding(12)
        .background(Color.orange)
        .cornerRadius(10)
    }

    private func resultsCard(result: VocalRangeResult) -> some View {
        VStack(spacing: 12) {
            HStack {
                VStack(alignment: .leading) {
                    Text("Low")
                        .font(.caption)
                        .foregroundColor(.white.opacity(0.7))
                    Text(result.low.noteLabel)
                        .font(.title)
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                    Text(String(format: "%.1f Hz", result.low.frequencyHz))
                        .font(.caption)
                        .foregroundColor(.white.opacity(0.7))
                }

                Spacer()

                VStack {
                    Text("Range")
                        .font(.caption)
                        .foregroundColor(.white.opacity(0.7))
                    Text(String(format: "%.1f oct", result.octaves))
                        .font(.title)
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                }

                Spacer()

                VStack(alignment: .trailing) {
                    Text("High")
                        .font(.caption)
                        .foregroundColor(.white.opacity(0.7))
                    Text(result.high.noteLabel)
                        .font(.title)
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                    Text(String(format: "%.1f Hz", result.high.frequencyHz))
                        .font(.caption)
                        .foregroundColor(.white.opacity(0.7))
                }
            }

            Divider()
                .background(Color.white.opacity(0.3))

            HStack {
                Text("Natural Shruti:")
                    .foregroundColor(.white.opacity(0.7))
                Text(result.naturalShruti.noteLabel)
                    .font(.title2)
                    .fontWeight(.bold)
                    .foregroundColor(.white)
            }
        }
        .padding(16)
        .background(Color.green)
        .cornerRadius(12)
    }

    private var controlButtons: some View {
        HStack {
            switch phase {
            case .idle:
                Button("Start Detection") {
                    start()
                }
                .buttonStyle(.borderedProminent)
                .frame(maxWidth: .infinity)

            case .complete, .cancelled:
                Button("Detect Again") {
                    reset()
                    start()
                }
                .buttonStyle(.borderedProminent)
                .frame(maxWidth: .infinity)

            default:
                Button("Cancel") {
                    stop()
                }
                .buttonStyle(.borderedProminent)
                .tint(.red)
                .frame(maxWidth: .infinity)
            }
        }
    }

    private func start() {
        // Create session with custom config (3-second countdown)
        let config = VocalRangeSessionConfig.custom(countdownSeconds: 3)
        let newSession = VocalRangeSession.create(config: config)
        session = newSession

        // Create recorder
        let tempPath = FileManager.default.temporaryDirectory
            .appendingPathComponent("range_temp.m4a").path
        recorder = SonixRecorder.create(outputPath: tempPath, format: "m4a", quality: "voice")

        guard let recorder = recorder else { return }

        // Observe session state via AsyncSequence
        observerTask = Task {
            for await state in newSession.state {
                await MainActor.run {
                    self.phase = state.phase
                    self.countdownSeconds = state.countdownSeconds
                    self.phaseMessage = state.phaseMessage
                    self.currentPitch = state.currentPitch
                    self.currentAmplitude = state.currentAmplitude
                    self.stabilityProgress = state.stabilityProgress
                    self.bestLowNote = state.bestLowNote
                    self.bestHighNote = state.bestHighNote
                    self.lowNote = state.lowNote
                    self.highNote = state.highNote
                    self.result = state.result
                    self.error = state.error
                }
            }
        }

        // Start session (runs auto-flow)
        newSession.start()

        // Start recording
        recorder.start()

        // Feed audio to session - resample to 16kHz first
        audioTask = Task {
            let hwRate = Int(Sonix.hardwareSampleRate)
            for await buffer in recorder.audioBuffers {
                // Resample to 16kHz for Calibra (expects 16kHz input)
                let samples16k = SonixResampler.resample(
                    samples: buffer.floatSamples,
                    fromRate: hwRate,
                    toRate: 16000
                )
                newSession.addAudio(samples: samples16k)
            }
        }
    }

    private func stop() {
        audioTask?.cancel()
        audioTask = nil
        observerTask?.cancel()
        observerTask = nil
        recorder?.stop()
        session?.cancel()

        // Reset UI state that might persist after cancel
        countdownSeconds = 0
        phase = .cancelled
    }

    private func reset() {
        stop()
        // Reset state primitives
        phase = .idle
        countdownSeconds = 0
        phaseMessage = "Ready to detect your vocal range"
        currentPitch = nil
        currentAmplitude = 0
        stabilityProgress = 0
        bestLowNote = nil
        bestHighNote = nil
        lowNote = nil
        highNote = nil
        result = nil
        error = nil
    }

    private func cleanup() {
        stop()
        recorder?.release()
        recorder = nil
        session?.release()
        session = nil
    }
}

private struct PhaseIndicator: View {
    let label: String
    let isActive: Bool
    let isComplete: Bool

    var body: some View {
        VStack(spacing: 4) {
            ZStack {
                Circle()
                    .fill(backgroundColor)
                    .frame(width: 28, height: 28)
                if isComplete {
                    Text("✓")
                        .font(.caption)
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                }
            }
            Text(label)
                .font(.caption)
                .foregroundColor(isActive ? .blue : .secondary)
        }
    }

    private var backgroundColor: Color {
        if isActive {
            return .blue
        } else if isComplete {
            return .green
        } else {
            return Color(.secondarySystemBackground)
        }
    }
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/calibra/VocalRangeSection.kt
package com.musicmuni.voxatrace.demo

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.musicmuni.voxatrace.calibra.VocalRangePhase
import com.musicmuni.voxatrace.calibra.VocalRangeSession
import com.musicmuni.voxatrace.calibra.VocalRangeState
import com.musicmuni.voxatrace.sonix.Sonix
import com.musicmuni.voxatrace.sonix.SonixRecorder
import com.musicmuni.voxatrace.sonix.SonixResampler
import kotlinx.coroutines.launch

/**
 * Vocal Range Detector using VocalRangeSession API.
 *
 * Demonstrates the simplified high-level API with observable state.
 */
@Composable
fun VocalRangeSection() {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    var session by remember { mutableStateOf<VocalRangeSession?>(null) }
    var recorder by remember { mutableStateOf<SonixRecorder?>(null) }
    val state by session?.state?.collectAsState() ?: remember { mutableStateOf(VocalRangeState()) }

    fun start() {
        // Create session and recorder
        session = VocalRangeSession.create()
        recorder = SonixRecorder.create("${context.cacheDir}/range_temp.m4a", "m4a", "voice")

        // Start session (runs auto-flow)
        session?.start()

        // Start recording and feed audio to session
        recorder?.start()
        scope.launch {
            val hwRate = Sonix.hardwareSampleRate
            recorder?.audioBuffers?.collect { buffer ->
                // Resample to 16kHz and convert to ShortArray
                val resampled = SonixResampler.resample(buffer.floatSamples, hwRate, 16000)
                val samples = ShortArray(resampled.size) { i ->
                    (resampled[i] * 32767).toInt().coerceIn(-32768, 32767).toShort()
                }
                session?.addAudio(samples)
            }
        }
    }

    fun stop() {
        recorder?.stop()
        session?.cancel()
    }

    fun reset() {
        recorder?.stop()
        session?.reset()
    }

    DisposableEffect(Unit) {
        onDispose {
            recorder?.release()
            session?.release()
        }
    }

    // UI
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Text("Vocal Range Detector", style = MaterialTheme.typography.titleMedium)

        Text(
            text = state.error ?: state.phaseMessage,
            style = MaterialTheme.typography.bodyMedium,
            color = if (state.error != null) MaterialTheme.colorScheme.error
                    else MaterialTheme.colorScheme.onSurfaceVariant
        )

        // Phase indicators
        PhaseIndicators(phase = state.phase)

        // Countdown
        if (state.phase == VocalRangePhase.COUNTDOWN && state.countdownSeconds > 0) {
            Text(
                text = "${state.countdownSeconds}",
                style = MaterialTheme.typography.displayLarge,
                modifier = Modifier.fillMaxWidth(),
                textAlign = TextAlign.Center
            )
        }

        // Real-time pitch display during detection
        if (state.phase == VocalRangePhase.DETECTING_LOW || state.phase == VocalRangePhase.DETECTING_HIGH) {
            PitchDisplayCard(
                noteLabel = state.currentPitch?.noteLabel ?: "-",
                frequencyHz = state.currentPitch?.frequencyHz ?: 0f,
                isStable = state.stabilityProgress >= 1.0f
            )

            // Stability progress
            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text(
                    text = "Stability: Hold for 1 second",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                LinearProgressIndicator(
                    progress = { state.stabilityProgress.coerceIn(0f, 1f) },
                    modifier = Modifier.fillMaxWidth().height(8.dp),
                    color = if (state.stabilityProgress >= 1.0f) Color(0xFF4CAF50)
                            else MaterialTheme.colorScheme.primary
                )
            }

            // Level meter
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text("Level:", style = MaterialTheme.typography.bodySmall)
                LinearProgressIndicator(
                    progress = { state.currentAmplitude.coerceIn(0f, 1f) },
                    modifier = Modifier.weight(1f).height(8.dp)
                )
            }
        }

        // Results
        state.result?.let { result ->
            ResultsCard(
                lowLabel = result.low.noteLabel,
                lowHz = result.low.frequencyHz,
                highLabel = result.high.noteLabel,
                highHz = result.high.frequencyHz,
                octaves = result.octaves,
                naturalShruti = result.naturalShruti.noteLabel
            )
        }

        // Controls
        Row(modifier = Modifier.fillMaxWidth()) {
            when (state.phase) {
                VocalRangePhase.IDLE -> {
                    Button(onClick = { start() }, modifier = Modifier.weight(1f)) {
                        Text("Start Detection")
                    }
                }
                VocalRangePhase.COMPLETE, VocalRangePhase.CANCELLED -> {
                    Button(onClick = { reset(); start() }, modifier = Modifier.weight(1f)) {
                        Text("Detect Again")
                    }
                }
                else -> {
                    Button(
                        onClick = { stop() },
                        modifier = Modifier.weight(1f),
                        colors = ButtonDefaults.buttonColors(containerColor = MaterialTheme.colorScheme.error)
                    ) {
                        Text("Cancel")
                    }
                }
            }
        }
    }
}

@Composable
private fun PhaseIndicators(phase: VocalRangePhase) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceEvenly
    ) {
        PhaseIndicator("Ready", phase == VocalRangePhase.COUNTDOWN, phase.ordinal > VocalRangePhase.COUNTDOWN.ordinal)
        Text("→", modifier = Modifier.padding(top = 8.dp))
        PhaseIndicator("Low", phase == VocalRangePhase.DETECTING_LOW, phase.ordinal > VocalRangePhase.DETECTING_LOW.ordinal)
        Text("→", modifier = Modifier.padding(top = 8.dp))
        PhaseIndicator("High", phase == VocalRangePhase.DETECTING_HIGH, phase == VocalRangePhase.COMPLETE)
        Text("→", modifier = Modifier.padding(top = 8.dp))
        PhaseIndicator("Done", phase == VocalRangePhase.COMPLETE, false)
    }
}

@Composable
private fun PhaseIndicator(label: String, isActive: Boolean, isComplete: Boolean) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Box(
            modifier = Modifier
                .size(28.dp)
                .background(
                    when {
                        isActive -> MaterialTheme.colorScheme.primary
                        isComplete -> Color(0xFF4CAF50)
                        else -> MaterialTheme.colorScheme.surfaceVariant
                    },
                    CircleShape
                ),
            contentAlignment = Alignment.Center
        ) {
            if (isComplete) {
                Text("✓", color = Color.White, style = MaterialTheme.typography.labelSmall)
            }
        }
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = if (isActive) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
private fun PitchDisplayCard(noteLabel: String, frequencyHz: Float, isStable: Boolean) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = if (isStable) Color(0xFF4CAF50).copy(alpha = 0.3f)
                             else MaterialTheme.colorScheme.primaryContainer
        )
    ) {
        Column(
            modifier = Modifier.fillMaxWidth().padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text(noteLabel, style = MaterialTheme.typography.displayMedium)
            Text("${"%.1f".format(frequencyHz)} Hz", style = MaterialTheme.typography.bodyLarge)
        }
    }
}

@Composable
private fun ResultsCard(
    lowLabel: String, lowHz: Float,
    highLabel: String, highHz: Float,
    octaves: Float, naturalShruti: String
) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(containerColor = Color(0xFF4CAF50))
    ) {
        Column(
            modifier = Modifier.fillMaxWidth().padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Column {
                    Text("Low", style = MaterialTheme.typography.labelMedium, color = Color.White.copy(alpha = 0.7f))
                    Text(lowLabel, style = MaterialTheme.typography.headlineMedium, color = Color.White)
                    Text("${"%.1f".format(lowHz)} Hz", style = MaterialTheme.typography.bodySmall, color = Color.White.copy(alpha = 0.7f))
                }
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text("Range", style = MaterialTheme.typography.labelMedium, color = Color.White.copy(alpha = 0.7f))
                    Text("${"%.1f".format(octaves)} oct", style = MaterialTheme.typography.headlineMedium, color = Color.White)
                }
                Column(horizontalAlignment = Alignment.End) {
                    Text("High", style = MaterialTheme.typography.labelMedium, color = Color.White.copy(alpha = 0.7f))
                    Text(highLabel, style = MaterialTheme.typography.headlineMedium, color = Color.White)
                    Text("${"%.1f".format(highHz)} Hz", style = MaterialTheme.typography.bodySmall, color = Color.White.copy(alpha = 0.7f))
                }
            }
            HorizontalDivider(color = Color.White.copy(alpha = 0.3f))
            Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.Center) {
                Text("Natural Shruti: ", color = Color.White.copy(alpha = 0.7f))
                Text(naturalShruti, style = MaterialTheme.typography.titleLarge, color = Color.White)
            }
        }
    }
}
